src/main/java/org/skriptlang/skript/lang/function/Contract.java

```diff
@@ -0,0 +1,44 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package org.skriptlang.skript.lang.function;
+
+import ch.njol.skript.lang.Expression;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The 'contract' of a function or another callable.
+ * This is a non-exhaustive helper for type hints, singularity, etc. that may change based on the arguments
+ * passed to a callable, in order for it to make better judgements on correct use at parse time.
+ */
+public interface Contract {
```

Comment:
Actually yes! I separated this because I felt it was something we might want to use in other fun places (like if we ever made another kind of functional _thing_, like a lambda or a subroutine section or some kind) or if it might pop up in something like skript-reflect too, since it's now our responsibility to take care of that.
Generally I make an effort to abstract any architecture I can so that we can reuse it in future as needed :)

Reply:
It might be better to stick this in a util package then